arch/xtensa/src/esp32/esp32_himem_chardev.c

```diff
@@ -0,0 +1,414 @@
+      priv->offset = offset;
+      break;
+    case SEEK_CUR:
+      priv->offset += offset;
+      break;
+    case SEEK_END:
+      priv->offset = dev->size + offset;
+      break;
+    default:
+      _err("invalid parameter: whence:%d\n", whence);
+      nxsem_post(&s_sem);
+      return -1;
+  }
+  filep->f_pos = priv->offset;
+  nxsem_post(&s_sem);
+  return priv->offset;
+}
+
+static int himem_chardev_ioctl(struct file *filep,
+                                int cmd, unsigned long arg)
+{
+  return 0;
+}
+
+static const struct file_operations fops =
+{
+  .open = himem_chardev_open,
+  .close = himem_chardev_close,
+  .read = himem_chardev_read,
+  .write = himem_chardev_write,
+  .seek = himem_chardev_seek,
+  .ioctl = himem_chardev_ioctl,
+};
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+int himem_chardev_init(void)
+{
+  int ret = 0;
+  ret = esp_himem_init();
+  if (ret != 0)
+    {
+      _err("Failed to himem init.\n");
+      return ret;
+    }
+
+  ret = esp_himem_alloc_map_range(ESP_HIMEM_BLKSZ, &s_range_handle); /* 32KB */
+  if (ret != 0)
+    {
+      _err("Failed to alloc himem map.\n");
+      return ret;
+    }
+
+  ret = nxsem_init(&s_sem, 0, 1);
+  if (ret != 0)
+    {
+      _err("Failed to init semaphore.\n");
+      esp_himem_free_map_range(s_range_handle);
+      return ret;
+    }
+
+  s_ram_offset = HIMEM_UNMAPPED;
+  s_mapped_filep = NULL;
+  return ret;
+}
+
+int himem_chardev_exit(void)
+{
+  int ret = 0;
+  ret = nxsem_destroy(&s_sem);
+  if (ret != 0)
+    {
+      _err("Failed to destroy semaphore.\n");
+      esp_himem_free_map_range(s_range_handle);
+      return ret;
+    }
+
+  ret = esp_himem_free_map_range(s_range_handle);
+  return ret;
+}
+
+int himem_chardev_register(char *name, size_t size)
+{
+  int ret = 0;
+  struct himem_chardev *dev;
+  dev = (struct himem_chardev *)malloc(sizeof(struct himem_chardev));
+  if (dev == NULL)
+    {
+      _err("Failed to malloc.\n");
+      return -1;
+    }
+
+  /* 32KB Alignment */
+
+  size_t mod = size % ESP_HIMEM_BLKSZ;
+  if (mod != 0)
+    {
+      size += (ESP_HIMEM_BLKSZ - mod);
+    }
+
+  dev->size = size;
+
+  strncpy(dev->name, name, 32);
+  ret = esp_himem_alloc(dev->size, &dev->mem_handle);
+  if (ret != 0)
+    {
+      _err("Failed to alloc himem. size=%d\n", dev->size);
+      free(dev);
+      return ret;
+    }
+
+  ret = register_driver(dev->name, &fops, 0666, dev);
+  if (ret != 0)
+    {
+      _err("Failed to register driver. dev=%s\n", dev->name);
+      esp_himem_free(dev->mem_handle);
+      free(dev);
+      return ret;
+    }
+
+  ret = nxsem_wait(&s_sem);
+  if (ret != 0)
+    {
+      _err("Failed to lock semaphore.\n");
+      unregister_driver(dev->name);
+      esp_himem_free(dev->mem_handle);
+      free(dev);
+      return ret;
+    }
+
+  list_add_tail(&s_himem_chardev_list, &dev->node);
```

Comment:
why add to a global list

Reply:
Because it is necessary to dynamically record and store the information of multiple character devices